### PR TITLE
feat(NODE-2843): implement sessions advanceClusterTime method

### PR DESCRIPTION
**Pull request by dariakp**
## Description

NODE-2843

**What changed?**

Implemented the advanceClusterTime session method, refactored internal helper function for naming consistency

### CODE REVIEW

File: test/unit/core/sessions.test.js
Comment (nbbeeken) [inline test/unit/core/sessions.test.js:4]:
```suggestion
const BSON = require('../../../src/bson');
```
Prefer importing bson from the driver so that the BSON-EXT logic (or anything else) is tested automatically.

File: src/sdam/common.ts
Comment (nbbeeken) [inline src/sdam/common.ts:70]: Just a thought, we could just keep this without the underscore and put `@internal` in the tsdoc?

Comment (dariakp) [reply src/sdam/common.ts:70]: I don't mind renaming, but this particular suggestion would make the name clash with the actual public method, so maybe we can come up with something else?

File: test/unit/core/sessions.test.js
Comment (dariakp) [reply test/unit/core/sessions.test.js:4]: good call out, though all test files except for the bson import one require it directly - any thoughts on that?

Comment (nbbeeken) [reply test/unit/core/sessions.test.js:4]: Hm, seems like something we should fix. In test/.eslintrc.json you can add bson to the forbidden modules, but there's 24 instances and at least two are intentionally testing bson vs bson-ext, so I guess lets defer to another ticket?

File: src/sdam/common.ts
Comment (nbbeeken) [reply src/sdam/common.ts:70]: I think the clash is fine because the public one is a method so it can't be called without a session. But maybe: `advanceClusterTimeFor` to imply its for the topology or session... I mean I'm splitting hairs, I just don't love underscore names but I'm not going to hold us up on this 😆

Comment (dariakp) [reply src/sdam/common.ts:70]: Yeah I understand that it's technically fine to use the same name, I just don't like to as a general practice because it makes it harder to tell at a glance (when doing a codebase wide search, for example) whether or not something is referencing the same thing. I don't have objections to something like `advanceClusterTimeFor` or `advanceClusterTimeOn` etc

File: test/unit/core/sessions.test.js
Comment (dariakp) [reply test/unit/core/sessions.test.js:4]: yeah, feels a bit out of scope, can file a simple clean up ticket in the test epic; would you rather I do use the internal reference for this test though?

File: src/sdam/common.ts
Comment (nbbeeken) [reply src/sdam/common.ts:70]: Both of those sound good to me, I leave the decision to you :)

File: test/unit/core/sessions.test.js
Comment (nbbeeken) [reply test/unit/core/sessions.test.js:4]: Totally, I don't imagine there being any impact here so whichever for now and we'll clean them up in that ticket.

File: src/sdam/common.ts
Comment (nbbeeken) [reply src/sdam/common.ts:70]: Decided to go with underscore

**Review by nbbeeken — APPROVED** [review]
LGTM

File: src/sessions.ts
Comment (dariakp) [inline src/sessions.ts:270]: @nbbeeken Just bringing your attention to this change

Comment (nbbeeken) [reply src/sessions.ts:270]: Just so I do not block this PR: looks correct to me.

Remind me to tell you the deeper issue here / file a ticket next week, has to do with BSON `promote*` options I think.

**Review by durran — APPROVED** [review]

**Pull request merged**
